[DNM] String interpolation rework

beccadax opened the pull request:
Implements [apple/swift-evolution#904](https://github.com/apple/swift-evolution/pull/904).

This is the implementation of the [string interpolation rework](https://forums.swift.org/t/draft-fix-string-interpolation-swift-5-edition/14786). It is now feature-complete, but there are still a few things on my to-do list:

* ~~It's performing similarly to the current design overall, although individual benchmarks vary wildly—from 1.80x for StringInterpolationSmall to 0.59x for FloatingPointPrinting_Float_interpolated. (Some of the improvements in tests which have little to do with string interpolation are probably thanks to [new overloads of `String.init(describing:)`](https://github.com/apple/swift/blob/8f41627c4c497e38f04c81ed0758cde403bf19a7/stdlib/public/core/Mirror.swift#L624-L760) I added to preserve source compatibility.) We want it to be faster than the status quo. I haven't dug deeply into this yet, and I'd like a second opinion from the CI.~~

* I'm wondering if it would be useful for [`DefaultStringInterpolation`](https://github.com/apple/swift/blob/8f41627c4c497e38f04c81ed0758cde403bf19a7/stdlib/public/core/StringInterpolation.swift#L59-L60) to become `moveonly` in the future, and if so, if that could be done without breaking ABI compatibility. @rjmccall, what do you think?

* There are a couple of test failures I'm not sure what to do with:

  * ~~Constraints/rdar42678836.swift was apparently added three days ago and has already had its message change twice. @slavapestov, @xedin, can we pin down what this test ought to do?~~

  * IDE/complete_at_top_level.swift fails because we don't suggest a Void function in a string interpolation. I *think* that comes from [this line](https://github.com/apple/swift/blob/master/lib/IDE/CodeCompletion.cpp#L953), which looks like it's intended to prevent you from completing a Void function where a value is expected, but I'm just guessing at its purpose. Can anyone clear up my confusion and help me figure out what this should be doing? (@nkcsgexi wrote this, but it was years ago.)

@swift-ci Please benchmark

cc @milseman

beccadax committed:
Tweak TypeChecker::callWitness() signature

It doesn’t actually need its array to be mutable.

beccadax committed:
Character->String through a less circular path

beccadax committed:
Extract NSCoding unstable name diagnosis

We’ll want to add more code near this, so it’s threatening to get unweildy.

beccadax committed:
Add TapExpr

This is not used for anything yet, but it will contain the append* calls on the StringInterpolation instance.

beccadax committed:
Permit empty interpolations

As in, “foo\()baz”. These were previously hard syntax errors due to a quirk in Parser::parseList(); there’s no particular reason they should be.

beccadax committed:
Refactor getCalledValue()

To also support a not-yet-used getCalledDeclRef().

beccadax committed:
Rework ExpressibleByStringInterpolation

beccadax committed:
Rip out typeCheckExpressionShallow()

It is now dead code.

beccadax committed:
Tweak fix-it locations for throwing interpolators

beccadax committed:
Diagnose missing appendInterpolation methods

Since we can’t declare a formal appendInterpolation(…) requirement, we manually check that there is at least one method which:

* is as public as the type itself,
* returns Void or has an @discardableResult, and
* is not static.

If there isn’t, we emit an error on the conformance and a warning for each appendInterpolation method present, indicating why it is not suitable.

beccadax committed:
Warn about strange interpolations in Swift 4.2

beccadax committed:
Late-breaking string interpolation test fixes

brentdax commented:
Don't think the bot actually saw that.

@swift-ci Please benchmark

beccadax committed:
Conform float types to TextOutputStreamable

This should improve interpolation performance, but I don’t know what impact the refactoring might have on .description, so we’ll see if we keep it.

brentdax commented:
Good thing I asked the CI—this is much closer to what I expected. [StringInterpolation is expected to be mediocre](https://gist.github.com/brentdax/fe23d4d9b0fdf8cd1949094f6ef1a936#implementation-details), so the only surprise is that the floating-point benchmarks aren't great.

To address that, I've added a `TextOutputStreamable` conformance to our floating-point types. I've also added an underscored fast path through `TextOutputStreamable` for buffers of ASCII characters. This improved the interpolation tests 1.3x on my machine—we'll see if that holds up.

@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci Please benchmark

beccadax committed:
Remove @_effects(readonly) helper functions

New string interpolation design can’t use @_effects(readonly), so these are no longer useful.

slavapestov commented on include/swift/AST/Expr.h:
```diff
@@ -938,6 +938,41 @@ class StringLiteralExpr : public LiteralExpr {
+  BraceStmt *Body;
+  
+public:
+  TapExpr(Expr * SubExpr, BraceStmt *Body);
```
Style: No space after *

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -4581,6 +4581,18 @@ namespace {
+      auto type = simplifyType(cs.getType(E));
+
+      E->getVar()->setType(type);
+      E->getVar()->setInterfaceType(type);
```
type->mapTypeOutOfContext()

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -7766,6 +7778,7 @@ namespace {
   class ExprWalker : public ASTWalker {
     ExprRewriter &Rewriter;
     SmallVector<ClosureExpr *, 4> ClosuresToTypeCheck;
+    SmallVector<std::tuple<TapExpr *, DeclContext *>, 4> TapsToTypeCheck;
```
Minor: we tend to use std::pair if there are two, but it doesn't matter

slavapestov commented on lib/Parse/Parser.cpp:
```diff
@@ -900,6 +900,10 @@ ParserStatus
 Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
                   bool AllowSepAfterLast, Diag<> ErrorDiag, SyntaxKind Kind,
                   llvm::function_ref<ParserStatus()> callback) {
+  auto TokIsStringInterpolationEOF = [&]() -> bool {
```
This needs a test

slavapestov commented on lib/AST/Expr.cpp:
```diff
@@ -1422,26 +1422,36 @@ DictionaryExpr *DictionaryExpr::create(ASTContext &C, SourceLoc LBracketLoc,
 
-static ValueDecl *getCalledValue(Expr *E) {
+template<typename T>
+static T getFromCalledDeclRefExpr(Expr *E, llvm::function_ref<T(const DeclRefExpr *)> DREFn,
```
I think these lines are too long?

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -2274,91 +2274,97 @@ namespace {
+        
+        Expr *base = TypeExpr::createImplicitHack(loc, type,
+                                                  tc.Context);
+        Expr *semanticExpr = tc.callWitness(base, dc, proto, *conformance,
```
Does this still rely on overload resolution or is it always calling the actual witness for the requirement?

brentdax replied:
This code (which generates the `init(literalCapacity:interpolationCount:)` and `init(stringInterpolation:)` calls) ought to always call the actual witness. That is what `TypeChecker::callWitness()` does, right? It does solve a constraint system, but I think it only uses it to calculate argument conversions, not to look for overloads.

Calls to `appendInterpolation` methods do use overload resolution; this is intended and necessary for certain desirable features and optimizations. Calls to the `appendLiteral(_:)` method also use overload resolution; this isn't really intended, but it'd be difficult to treat them differently from `appendInterpolation(_:)` without a more specialized AST representation.

slavapestov replied:
As long as the overloaded call is built at parse time or CSGen and resolved by the time we get to CSApply it’s fine.

callWitness() is overkill for what it does — we don’t need to create a constraint system to look up a single member and build a call. But as long as you’re not relying on it for overload resolution I’ll be able to clean it up later. My plan is to get rid of callWitness() and any other places in CSApply where we solve new constraint systems. I think typeCheckExpressionShallow() was the other main case but you removed that which is great.

slavapestov commented on lib/AST/Expr.cpp:
```diff
@@ -2166,6 +2182,55 @@ void KeyPathExpr::Component::setSubscriptIndexHashableConformances(
+          name = unresolvedDot->getName();
+        }
+        
+        segment.isInterpolation = (name.getBaseName() == "appendInterpolation");
```
Can you instead add this identifier and others to Identifiers.def so that it appears as a member of the ASTContext?

brentdax replied:
This identifier actually is in Identifiers.def. I don't use it here because I don't think I can reliably get to an `ASTContext` from this node, and I didn't want to pass a context parameter just to get that identifier. Should I just bite the bullet and do it anyway?

beccadax committed:
Apply feedback from @slavapestov

beccadax committed:
Simplify InterpolatedStringLiteralExpr::forEachSegment()

Removes the helper InterpolatedStringLiteralExpr::SegmentInfo struct. This also allows us to bring back the old ApplyExpr::getCalledValue() implementation.

beccadax committed:
Fix issues from rebasing

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please benchmark

beccadax committed:
Update for DeclContext renaming

brentdax commented:
Ugh, forgot to actually commit the rename...

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please benchmark

ravikandhadai commented:
@brentdax  This is about the default argument support we were discussing in the pitches thread. 
 I tried this PR (interpolation-rework) but I couldn't get the default arguments working.  For example,

    struct MyInterpolation: ExpressibleByStringInterpolation, StringInterpolationProtocol {
      mutating func appendInterpolation(_ num: Int, _ otherParam: Bool = false) {
      }
      ...
    }
    let x: MyInterpolation = "\(10)"

produces a compilation **error:** **value of type 'MyInterpolation.StringInterpolation' (aka 'MyInterpolation') has no member 'appendInterpolation'**

brentdax commented:
@ravikandhadai Thanks for letting me know. Can you provide a complete conformance that doesn't work? i.e. including the `init(stringInterpolation:)`, `init(literalCapacity:interpolationCount:)`, and `appendLiteral(_:)` methods. If you're not sure how to share your sample, [Github Gist](https://gist.github.com) is a reasonable choice.

slavapestov commented:
@brendax Default argument support might need some messing around with locators. Search CSSimplify.cpp for getCalleeDeclAndArgs()

ravikandhadai commented:
@brentdax  Here is a toy example with all conformances: https://gist.github.com/ravikandhadai/8c3d7cc0b2ebb264ef3c3e63901958e8

brentdax commented:
@ravikandhadai Thank you so much for the test case. It looks like you're right—for this example, the parser was forming a name like `appendInterpolation(_:)`, which would keep the type checker from trying other variants of `appendInterpolation` with default arguments. I've already manually verified that a one-line fix corrects this issue, but I'll add a regression test for this and push it up soon (probably tonight).

beccadax committed:
Fix method call on null object

brentdax commented:
@ravikandhadai I'm stopping for the night because fixing this broke something else, but I'll keep working on it soon.

ravikandhadai commented:
Sounds great. Thx.

beccadax committed:
Make both same-file extensions and default args work

Previously, the parser generated compound method names for appendInterpolation(…) methods because this helped when finding appendInterpolation methods declared in the same file. However, this turned out to prevent default arguments from working.

This commit returns it to adding base names only and instead explicitly calls loadAllMembers() on the StringInterpolation type. I’m not sure why types in other expressions don’t need this but types in these generated expressions do, but it’s much closer to the problem and doesn’t seem to have any ill effects.

brentdax commented:
@swift-ci please test source compatibility

brentdax commented:
@swift-ci please test source compatibility

brentdax commented:
The debug source compatibility failure doesn't happen locally, so I'm going to assume there's nothing fundamentally wrong and we can schedule a review on the proposal.

benrimmington commented on stdlib/public/core/StringInterpolation.swift:
```diff
@@ -10,11 +10,84 @@
+///     let price = 2
+///     let number = 3
+///     let message = "If one cookie costs \(price) dollars, " +
+///                   "\(number) cookies cost \(price * number) dollars."
```
This example was written more than two years ago (before [SE-0168][] and [SE-0182][]).

Would it be any better with a multi-line string?

```swift
///     let message = """
///                   If one cookie costs \(price) dollars, \
///                   \(number) cookies cost \(price * number) dollars.
///                   """
```

Unless there is a compile-time optimization for the concatenation?

[SE-0168]: <https://github.com/apple/swift-evolution/blob/master/proposals/0168-multi-line-string-literals.md>
[SE-0182]: <https://github.com/apple/swift-evolution/blob/master/proposals/0182-newline-escape-in-strings.md>

brentdax replied:
I'll change these throughout.

beccadax committed:
[NFC] Improve debuggability of lookupDirect()

This change:

1. Adds accessors for the bit attached to NominalTypeDecl::LookupTable. These let you more easily search for and break on changes; more importantly, they give this bit a name and some documentation explaining what (I think) it means.

2. Includes the “ignoreNewExtensions” parameter in the debug output from NominalTypeDecl::lookupDirect().

3. Adds a MemberLookupTable::dump() method.

beccadax committed:
[NFC] Remove dead MemberLookupTable::addExtensionMembers()

These methods are never used and appear to be holdovers from an earlier implementation.

beccadax committed:
[AST] Update lazy LookupTable when extension added

Previously, nothing would guarantee that, if a nominal type `T` with lazy members had extensions adding members with name `foo`, and these members were already in `T`’s LookupTable, and a new extension to `T` added another member named `foo`, the new extension’s member would be added to the LookupTable.

This change makes it so that adding an extension clears the isLookupTablePopulated() flag, and so that when the flag is cleared on a type with lazy members, it updates all existing entries instead of just the ones present in the type itself.

Finally, it removes a workaround in string interpolation which is no longer necessary because of this change.

brentdax commented:
I've traced the extension bug down further and, in 528acb1, I have a less magical fix. I'm not sure it's quite right, though, so I'd like it to be reviewed by someone who understands name lookup better than I do—possibly @graydon, who wrote this code.

What's happening is that, in code like this:

```swift
_ = "\("hello")"    // implicit call to DefaultStringInterpolation.appendInterpolation(_:)
extension DefaultStringInterpolation {
  mutating func appendInterpolation(foo: String) {}
}
_ = "\(foo: "hi")"    // implicit call to DefaultStringInterpolation.appendInterpolation(foo:)
```

Name resolution on the last line is finding the overloads of `appendInterpolation(_:)`, but not `appendInterpolation(foo:)`. This is happening because the entry for `appendInterpolation` in the `LookupTable` is generated while type-checking the call on the first line, and is not updated before or during type-checking of the call on the last line. (`DefaultStringInterpolation` comes from the standard library and has lazy members.)

I *think* the mistake is twofold:

1. `prepareLookupTable()` ought to update all of the entries in the lookup table to include newly-added extensions, not just the ones directly in the type.

2. The `LookupTable` is not marked as being out of date or needing any sort of rebuilding.

So I've made both of those changes in 528acb1. But it's possible I've misunderstood how this code is supposed to be structured and there's a better way to solve this.

I've also pushed 53b68a5, which adds some debugging affordances for name lookup and explicitly names the flag attached to `LookupTable`, and 528acb1, which removes some confusing dead code. I'll probably cherry-pick these two or possibly all three and send them as a separate pull request; I just want to make sure I'm on the right track first.

cc @slavapestov

DougGregor commented:
@swift-ci build toolchain

DougGregor commented:
@brentdax would you please rebase this? I'd like to build a toolchain

brentdax commented:
@DougGregor Will do.

brentdax commented:
@swift-ci build toolchain

brentdax commented:
@swift-ci build toolchain

beccadax committed:
Fix Linux-only build failure

Apparently, the macOS STL is more forgiving about tuples vs. pairs than the Linux one.

brentdax commented:
@swift-ci build toolchain Linux platform

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Benchmark interpolation wtih custom types

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Add missing @inlinable annotation

Probably responsible for at least part of performance gap.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Correct arguable CustomStringInterpolation benchmark flaw

As currently written, the optimizer can completely remove the involvement of CustomString in some cases. We probably don’t want that, so let’s fix it.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Test inlining an empty-string check

beccadax committed:
Inline the SmallString check for reserveCapacity()

brentdax commented:
@swift-ci smoke benchmark

brentdax commented:
@swift-ci smoke benchmark

brentdax commented:
I don't know if SE-0228 will be approved today, but if it is, this PR still needs a round of cleanup and a thorough hazing by the CI bot before I think it's ready to be reviewed and merged.

beccadax committed:
Update for function type changes

beccadax committed:
Update stable ABI for Digester tests

beccadax committed:
Update sil_location test for updated mangling

beccadax committed:
Change suspect code completion test

This change tests that, in string interpolation, code completion doesn’t suggest Void functions but suggests functions returning other types.

This is a change in behavior, but I’m not convinced the old version was correct in the first place—why would you want to interpolate a function that returns Void? It’s valid but not useful, and code completion seems to try to avoid suggesting Void functions in other similar circumstances.

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please smoke test

milseman commented on stdlib/public/core/StringInterpolation.swift:
```diff
@@ -26,48 +101,127 @@ extension String : _ExpressibleByStringInterpolation {
+  ///                   "\(number) cookies cost \(price * number) dollars."
+  ///     print(message)
+  ///     // Prints "If one cookie costs 2 dollars, 3 cookies cost 6 dollars."
+  @inlinable
```
Why is this `@inlinable`?

brentdax replied:
In general, these benchmarked faster when they were inlinable, and IIRC the `init(stringInterpolationSegment:)` overloads they replaces were inlinable/transparent. We can retest/reconsider this, though.

milseman commented on stdlib/public/core/StringProtocol.swift:
```diff
@@ -17,9 +17,10 @@
   Hashable, Comparable
-  where Iterator.Element == Character, SubSequence : StringProtocol {
+  where Iterator.Element == Character, SubSequence : StringProtocol,
+        StringInterpolation == DefaultStringInterpolation {
```
Out of curiosity, why is the new requirement needed? Do we want to make StringProtocol conform? Do we even want Substring to?

brentdax replied:
It's not strictly necessary to support string interpolation in `Substring`s. But it's slightly weird for users that we don't, and it's really easy to do so now that we have `DefaultStringInterpolation`.

It would probably make it difficult to handle `StringProtocol`-conforming types which aren't built around `String`, though. If we're still thinking we might want to do that, we would probably need this to be something like `DefaultStringInterpolation<Base>`, where `Base` was an associated type that would be `String` for both `String` and `Substring`.

milseman commented on test/Parse/try.swift:
```diff
@@ -180,3 +180,19 @@ func test(_: X) {}
+  mutating func appendInterpolation() throws {}
+}
+
+_ = try "a\()b"
```
Does this work with a mixture of throwing and non-throwing interpolation calls?

milseman reviewed (commented):
Mostly nitpicks

milseman commented:
@slavapestov and @ravikandhadai, do you think you can look at the compiler-side changes?

brentdax commented on benchmark/single-source/StringInterpolation.swift:
```diff
@@ -104,3 +120,49 @@ public func run_StringInterpolationManySmallSegments(_ N: Int) {
+      // FIXME: if String is not stored as UTF-16 on this platform, then the
+      // following operation has a non-trivial cost and needs to be replaced
+      // with an operation on the native storage type.
+      result = result &+ Int(utf16.last!)
```
I think that line is copied from the other benchmarks—if we want to change it, we should probably change all of them.

brentdax commented on stdlib/public/core/CompilerProtocols.swift:
```diff
@@ -703,49 +703,178 @@ public protocol ExpressibleByDictionaryLiteral {
+  ///                   "\(number) cookies cost \(price * number) dollars."
+  ///     print(message)
+  ///     // Prints "If one cookie costs 2 dollars, 3 cookies cost 6 dollars."
+  public init(stringInterpolation: DefaultStringInterpolation) {
```
This is a default implementation of a protocol requirement that users often should implement themselves. All this is really saying is “this exists for a specific purpose, and normally the compiler writes calls to it for you”. I believe `init(stringLiteral:)` has similar language.

brentdax commented on stdlib/public/core/StringGuts.swift:
```diff
@@ -910,19 +910,23 @@ extension _StringGuts {
+    reserveCapacitySlow(capacity)
+  }
+
+  @usableFromInline // @testable
```
We might actually want to pull cf9f30f and 91065fc out of this PR and consider them separately—they're basically `String` optimizations that happen to improve string interpolation performance.

beccadax committed:
Use multiline string literals in examples

beccadax committed:
Remove inaccurate comments

beccadax committed:
Improve "informal requirement" comment

beccadax committed:
Test mixture of throwing and nonthrowing interpolations

beccadax committed:
Style fixes

brentdax commented:
@swift-ci please smoke test

beccadax committed:
Delete ExpressibleByStringInterpolation deprecation test

brentdax commented:
The failing test on Linux means that it can't completely optimize away a string interpolation in a Void context. This was already true on some architectures ([SR-7806](https://bugs.swift.org/browse/SR-7806)), but it's now happening on 64-bit platforms too.

brentdax commented:
@swift-ci please test source compatibility

@swift-ci please smoke benchmark

@swift-ci please test compiler performance

brentdax commented:
@swift-ci please test source compatibility

brentdax commented:
The source compatibility debug failure is from Fluent, which seems to cause flaky failures.

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
It seems like certain fast-ish benchmarks slow down; I want to see how stable the set is.

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
XFAIL Linux-only test failure

This already didn't work on 32-bit platforms; now it's not working on 64-bit Linux. Filed as SR-9008.

brentdax commented:
@swift-ci please test

beccadax committed:
Change Equal constraints to Bind

This may improve type-checking performance somewhat.

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

brentdax commented:
@xedin Walltime's still higher, but not higher enough to make the "regressed" list.

xedin commented:
The way I see it, it's a win already, because we do a lot less work in the solver, number of scopes allocated is significantly lower. I think at this point the additional times comes from the `NumDeclsValidated`...

airspeedswift commented:
I'm wary of how on the smoke test is excluding so many projects right now: Kingfisher, ReactiveCocoa, ReactiveSwift. Not sure how many that leaves but these results might not be representative (in either direction). Maybe try non-smoke version. cc @graydon

airspeedswift commented:
@swift-ci please test compiler performance

airspeedswift commented:
But I agree with @xedin, if it weren't for that concern, I'd say those results suggest it's gtm.

xedin commented:
@airspeedswift These 3 are always excluded I think and full test is still running, so let's see...

airspeedswift commented:
> ## release brief
> <details open>
> <summary>Improved (1)</summary>
> 
> name | delta_pct
> ---: |  ---:
> Frontend.NumInstructionsExecuted  | -45.6% :white_check_mark:
> </details>

Uh...

xedin commented:
I'll try to build your branch and see why there are so much difference in `NumConstraintsConsideredForEdgeContraction`, that might not be a problem though, but it indicates that constraint systems have more constraints then before.

brentdax commented:
Continuing in #19963.
